cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -0,0 +1,284 @@
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/palantir/stacktrace"
+	"golang.org/x/crypto/ssh"
+
+	"alluxio.org/cli/env"
+	"alluxio.org/log"
+)
+
+type Result struct {
+	err error
+	msg string
+}
+
+type HostnamesFile struct {
+	Name      string
+	Mutex     *sync.Mutex
+	Once      *sync.Once
+	Hostnames []string
+}
+
+func NewHostnamesFile(name string) *HostnamesFile {
+	return &HostnamesFile{
+		Name:  name,
+		Mutex: &sync.Mutex{},
+		Once:  &sync.Once{},
+	}
+}
+
+// for each node, create a client and run
+func runCommand(command string, mode string) error {
+	// prepare client config, ssh port info
+	config, port, err := prepareCommand()
+	if err != nil {
+		return stacktrace.Propagate(err, "prepare command failed")
+	}
+
+	// get list of masters or workers, or both
+	hosts, err := getHostnames(mode)
+	if err != nil {
+		return stacktrace.Propagate(err, "cannot read host names")
+	}
+
+	// create wait group and channels
+	var wg sync.WaitGroup
+	results := make(chan Result)
+	for _, h := range hosts {
+		wg.Add(1)
+		host := h
+
+		go func() {
+			defer wg.Done()
+			// dial on target remote address with given host, config and ssh port
+			dialAddr := fmt.Sprintf("%s:%d", host, port)
+			conn, err := ssh.Dial("tcp", dialAddr, config)
+			if err != nil {
+				result := Result{
+					err: err,
+					msg: fmt.Sprintf("dial failed to %v", host),
+				}
+				results <- result
+			}
+			defer func(conn *ssh.Client) {
+				if err := conn.Close(); err != nil {
+					log.Logger.Infof("connection to %s closed, error: %s", host, err)
+				} else {
+					log.Logger.Infof("connection to %s closed", host)
+				}
+			}(conn)
+
+			// create and set up a session
+			session, err := conn.NewSession()
+			if err != nil {
+				result := Result{
+					err: err,
+					msg: fmt.Sprintf("cannot create session at %v.", host),
+				}
+				results <- result
+			}
+			defer func(session *ssh.Session) {
+				err := session.Close()
+				if err != nil && err != io.EOF {
+					log.Logger.Infof("session at %s closed, error: %s", host, err)
+				} else {
+					log.Logger.Infof("session at %s closed", host)
+				}
+			}(session)
+
+			session.Stdout = os.Stdout
+			session.Stderr = os.Stderr
+
+			// run command on the session, output errors
+			if err = session.Run(command); err != nil {
+				result := Result{
+					err: err,
+					msg: fmt.Sprintf("run command %v failed at %v", command, host),
+				}
+				results <- result
+			}
+
+			// if no errors, return nil
+			result := Result{
+				err: nil,
+				msg: "",
+			}
+			results <- result
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
+
+	return errorHandler(command, hosts, results)
+}
+
+func prepareCommand() (*ssh.ClientConfig, int, error) {
+	// get the current user
+	cu, err := user.Current()
+	if err != nil {
+		return &ssh.ClientConfig{}, -1,
+			stacktrace.Propagate(err, "cannot find current user")
```

Comment:
```suggestion
		return nil, 0, stacktrace.Propagate(err, "cannot find current user")
```
return the type's zero values by default. for `int` type, sometimes `0` is an expected return value (ex. exit code); only then would i return `-1`

although i don't recommend it for this specific scenario, it can be useful to used [named return values](https://go.dev/tour/basics/7). you don't need to specify every value in the return statement. the variables are initialized to their zero value (this is also why i recommend defaulting to the zero value), updated as needed in the function, and then only need to have a `return` statement